shared/util/quit-helper.desktop.js

```diff
@@ -55,8 +55,20 @@ export function executeActions (actions: Array<Action>) {
   }
 }
 
-export function setupExecuteActionsListener () {
+type appType = {
+  once: (name: string, callback:
+    (event: {preventDefault: () => void}) => void
+  ) => void
+}
+export function setupExecuteActionsListener (app: appType) {
   ipcMain.on('executeActions', (event, actions) => {
     executeActions(actions)
   })
+
+  // quit through dock. only listen once
+  app.once('before-quit', event => {
+    console.log('Quit through dock')
+    event.preventDefault()
+    quit()
```

Comment:
I think we should add a context for `{type: 'dock'}` or `{type: 'beforeQuitEvent'}`

and follow the same execute actions, this way we split what we want to do with how we do it.

So this would be `executeActions(quitOnContext({type: 'beforeQuitEvent'}))`. Then change quitOnContext to define what you should do